### DISCUSSION TITLE
cmssw_columnar.sh sandbox not working in htcondor workflow

### DISCUSSION THREAD

**Discussion by JulesVandenbroeck** [columnflow/columnflow#525] [Q&A]
Hi,

I require the cmssw_columnar.sh sandbox for pyROOT BDT purposes in my calibration task
. I have setup the sandbox and it works fine when running locally. However when running the task with `--workflow htcondor` venv setup does not work properly as the error `ModuleNotFoundError: No module named 'numpy'` is shown in the job log. I tried uninstalling the cmssw_columnar software saved in `data` as well as rerun cf.BundleCMSSWSandbox. I also placed the job log file in attachment.

Kind Regards,
Jules Vandenbroeck

[stdall_7To8.txt](https://github.com/user-attachments/files/16522801/stdall_7To8.txt)

**Comment by pkausw** [comment]
Hi @JulesVandenbroeck ,

sorry for the late reply! Do you still have this problem? Looking at the error log, I see that you have more errors even before the `No module named 'numpy'` error you mention above. I'm guessing this doesn't when you run the job locally, right?

Can you send us more information, i.e. the link to your code/a minimal example and the command you've used? Also, I'm not familiar with the architecture of your cluster. Is there a difference between the htcondor workers and your "normal" login nodes? I've noticed that python3 on the htcondor worker is python3.6, which shouldn't make a difference since you've setup python3.9 via micromamba. However, in CMSSW you also link a lot of files from cvfms afaik, so maybe something interfers destructively on the htcondor worker, but not on your local login node? If this is the case, asking your local admin team might also be a good idea.

Cheers,
Philip

**Reply by JulesVandenbroeck** [reply]
Hi Philip,

With the errors other than the `No module named 'numpy'` you mean the ones below, these also appear when running locally (but the task still runs). [Here](https://gitlab.cern.ch/ghentanalysis/columnflowanalysis/ttz) is a link to the gitlab repository of my analysis, this code will however only work on our cluster. I also wanted to confirm if the columnflow template is able to run any task with the cmssw_columnar.sh sandbox or not?

I don't know the exact difference between the htcondor workers and the login nodes, I will contact the t2b experts as well to confirm this and see if they have a solution.

Cheers,
Jules


```
/ada_mnt/ada/user/juvanden/columnflowProjects/ttz/data/software/cmssw/cmssw_columnar_9b3fd8a8/CMSSW_12_6_2/venvs/columnar/bin/python3: error while loading shared libraries: libpython3.9.so.1.0: cannot open shared object file: No such file or directory
/ada_mnt/ada/user/juvanden/columnflowProjects/ttz/data/software/cmssw/cmssw_columnar_9b3fd8a8/CMSSW_12_6_2/venvs/columnar/bin/python3: error while loading shared libraries: libpython3.9.so.1.0: cannot open shared object file: No such file or directory
/ada_mnt/ada/user/juvanden/columnflowProjects/ttz/data/software/cmssw/cmssw_columnar_9b3fd8a8/CMSSW_12_6_2/venvs/columnar/bin/python3: error while loading shared libraries: libpython3.9.so.1.0: cannot open shared object file: No such file or directory
```

**Reply by pkausw** [reply]
Ah ok, yes I was referring to the missing shared objects errors. But if you also see them locally and the code runs, this is weird.

Thanks for providing the link to your code base, though it's of course a little tough to reproduce the behavior if the code only runs on your cluster. Personally, I'm not aware of any ongoing analysis that relies too much on the cmssw environments - I'll aks around and will also try to run a couple of sanity checks asap.

**Reply by pkausw** [reply]
Hi @JulesVandenbroeck , I was able to reproduce a similar error (if not the same) on our cluster with a derivative of the `cmssw_columnar` sandbox. I think I traced it down to this line [here](https://github.com/columnflow/columnflow/blob/master/sandboxes/cmssw_columnar.sh#L26). Can you try to change this to `pip install -I ...` (i.e. add a `-I`) to the command and reinstall the sandbox? This fixed the issue on my end at least. I'm also creating an issue to keep track, thanks for spotting this!

**Reply by JulesVandenbroeck** [reply]
I tried to replace `-r` with `-I` in `cmssw_columnar.sh`, however when trying to reinstall the sandbox with `modules/columnflow/sandboxes/cmssw_columnar.sh reinstall` I get the following error:
```
ERROR: Invalid requirement: '/user/juvanden/columnflowProjects/ttz/modules/columnflow/sandboxes/columnar.txt': Expected package name at the start of dependency specifier
    /user/juvanden/columnflowProjects/ttz/modules/columnflow/sandboxes/columnar.txt
    ^
Hint: It looks like a path. The path does exist. The argument you provided (/user/juvanden/columnflowProjects/ttz/modules/columnflow/sandboxes/columnar.txt) appears to be a requirements file. If that is the case, use the '-r' flag to install the packages specified within it.
```

**Reply by pkausw** [reply]
Sorry, I didn't mean that you should _replace_ anything, you just need to add `-I`. Side note, `-r` tells pip to use a requirements file, so `-r` must be placed before the path to the .txt file

**Reply by JulesVandenbroeck** [reply]
Sorry for the late response,
I have added the `-I` to `pip install ...` and now I am able to submit jobs with `cmssw_columnar` sandbox. This still only works for files that are locally stored (gfal2 is not installed) but this was a problem I also had in the [previous discussion](https://github.com/columnflow/columnflow/discussions/442) but does not affect my workflow.

Thank you for the help!
